### DISCUSSION TITLE
Memory leak !?

### DISCUSSION THREAD

**Discussion by anandhkrishna** [tianocore/edk2#6483] [General]
why we didn't free the hob memory  [here](https://github.com/tianocore/edk2/blob/5158b598f7a64459ffdc7bcd688d9e573008b8f6/MdeModulePkg/Core/Dxe/Gcd/Gcd.c#L2847)  , since the Hobs are moved to system memory [here](https://github.com/tianocore/edk2/blob/5158b598f7a64459ffdc7bcd688d9e573008b8f6/MdeModulePkg/Core/Dxe/DxeMain/DxeMain.c#L359)

**Comment by r1chard-lyu** [comment]
I’m interested in this topic as well and look forward to seeing more discussion.

**Comment by os-d** [comment]
@anandhkrishna @r1chard-lyu , I believe the answer here is that this is not a memory leak because the HOB list itself is not covered in a memory allocation HOB, so DXE Core does not see it as allocated memory, but part of free memory. That is why it must be relocated and tracked as allocated memory in DXE, so that it is not stomped on. See https://github.com/tianocore/edk2/blob/bff50932c1288728079f64bb51a3e86cede3a168/MdeModulePkg/Core/Pei/Hob/Hob.c#L104-L116 and the comment here: https://github.com/tianocore/edk2/blob/5158b598f7a64459ffdc7bcd688d9e573008b8f6/MdeModulePkg/Core/Dxe/Gcd/Gcd.c#L2841-L2845.

**Comment by r1chard-lyu** [comment]
@os-d Thank you for the detailed explanation. I understand now!
